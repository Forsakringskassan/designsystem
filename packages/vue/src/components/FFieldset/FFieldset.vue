<script lang="ts">
import { type PropType, defineComponent, provide, useSlots, useTemplateRef } from "vue";
import { ElementIdService, type ValidityEvent, debounce } from "@fkui/logic";
import { renderSlotText, dispatchComponentValidityEvent, hasSlot } from "../../utils";
import { ComponentValidityEvent } from "../../types";
import { FIcon } from "../FIcon";
import { TranslationMixin } from "../../plugins";
import { tooltipAttachTo } from "../FTooltip";
import { labelClasses } from "./label-classes";
import { contentClasses } from "./content-classes";
import { injectionKeys } from "./use-fieldset";

function isEqual<T>(a: T[], b: T[]): boolean {
    if (a.length !== b.length) {
        return false;
    }
    return a.every((_, i) => a[i] === b[i]);
}

export default defineComponent({
    name: "FFieldset",
    components: {
        FIcon,
    },
    mixins: [TranslationMixin],
    props: {
        /**
         * The id for the fieldset id attribute.
         * If the prop is not set a random value will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
        /**
         * Name provided to child content as `sharedName` for optional usage (it will not be set on the fieldset element).
         * For radio inputs this is a shortcut to specify the shared name attribute at one place,
         * instead of repeatedly setting the name attribute on each radio input.
         */
        name: {
            type: String,
            required: false,
            default: undefined,
        },
        /**
         * The CSS classes for the label, description and error-message slot.
         */
        labelClass: {
            type: String,
            required: false,
            default: "",
        },
        /**
         * The CSS classes for the default slot.
         */
        contentClass: {
            type: String,
            required: false,
            default: "",
        },
        /**
         * Aligns underlying items horizontally.
         * Supported by radiobuttons and chip layout.
         */
        horizontal: {
            type: Boolean,
            required: false,
        },
        /**
         * Displays radio and checkbox content with chip layout.
         */
        chip: {
            type: Boolean,
            required: false,
            default: false,
        },
        /**
         * Displays a box with border around radiobuttons and checkboxes.
         */
        border: {
            type: Boolean,
            required: false,
        },
        /**
         * Sets visibility behaviour for details slot in selectable child items. By default details slot is not rendered.
         *
         * * `never` (default) - Never show item details.
         * - `when-selected` - Show item details when selected.
         * - `always` - Always show item details.
         */
        showDetails: {
            type: String as PropType<"never" | "when-selected" | "always">,
            default: "never",
            validator(value: string) {
                return ["never", "when-selected", "always"].includes(value);
            },
        },
    },
    setup(props) {
        const slots = useSlots();
        provide(injectionKeys.sharedName, props.name);
        provide(injectionKeys.showDetails, props.showDetails);
        provide(injectionKeys.getFieldsetLabelText, () => {
            return renderSlotText(slots.label);
        });
        provide(tooltipAttachTo, useTemplateRef("tooltipAttachTo"));
    },
    data() {
        return {
            validity: {
                validityMode: "INITIAL",
            } as ValidityEvent,
            descriptionClass: ["label__description"],
            formatDescriptionClass: ["label__description", "label__description--format"],
            validityElement: null as HTMLElement | null,
            dispatchObject: {} as unknown as ComponentValidityEvent,
            detail: {} as unknown as ValidityEvent,
            hasDocumentListener: false,
            legendKey: 1,
            oldMessage: "",
            children: new Array<HTMLInputElement>(),
            hasCheckbox: false,
            hasRadiobutton: false,
        };
    },
    computed: {
        hasError(): boolean {
            return this.validity.validityMode === "ERROR";
        },
        hasErrorMessageSlot() {
            return hasSlot(this, "error-message");
        },
        hasTooltipSlot(): boolean {
            // return hasSlot(this, "tooltip");
            return Boolean(this.$slots.tooltip);
        },
        hasDescriptionSlot(): boolean {
            return hasSlot(this, "description");
        },
        legendClass(): string[] {
            // If the tooltip is provided we should avoid adding label class to the top legend element.
            return this.hasTooltipSlot ? ["sr-only"] : this.groupLabelClass;
        },
        groupLabelClass(): string[] {
            return Array.from(labelClasses(this));
        },
        groupContentClass(): string[] {
            return Array.from(contentClasses(this));
        },
        classes(): Record<string, boolean> {
            const { hasRadiobutton, hasCheckbox, horizontal, chip, border } = this;
            return {
                "radio-button-group": hasRadiobutton,
                "radio-button-group--chip": chip && hasRadiobutton,
                "radio-button-group--horizontal": horizontal && hasRadiobutton,
                "radio-button-group--border": border && hasRadiobutton,
                "checkbox-group": hasCheckbox,
                "checkbox-group--chip": chip && hasCheckbox,
                "checkbox-group--horizontal": horizontal && hasCheckbox,
                "checkbox-group--border": border && hasCheckbox,
            };
        },
        checkedChildren(): HTMLInputElement[] {
            return this.children.filter((child) => child.checked);
        },
        debouncedUpdateChildren(): () => void {
            return debounce(this.updateCheckboxChildren.bind(this), 150);
        },
        checkboxCheckedScreenReaderText(): string {
            return this.checkedChildren.length === 1
                ? this.$t("fkui.checkbox-group.checkbox.checked", "Kryssruta kryssad")
                : this.$t("fkui.checkbox-group.checkbox.not.checked", "Kryssruta ej kryssad");
        },
        numberOfCheckboxesScreenReaderText(): string {
            return this.$t("fkui.checkbox-group.count", "Grupp med {{ count }} kryssrutor", {
                count: String(this.children.length),
            });
        },
        numberOfCheckedCheckboxesScreenText(): string {
            return this.$t("fkui.checkbox-group.checked", "{{ checked }} kryssad av {{ count }}", {
                checked: String(this.checkedChildren.length),
                count: String(this.children.length),
            });
        },
    },
    async mounted(): Promise<void> {
        await this.$nextTick();
        const types = Array.from(
            this.$el.querySelectorAll(`input[type="checkbox"], input[type="radio"]`),
            (it: HTMLInputElement) => it.getAttribute("type"),
        );
        this.hasCheckbox = types.includes("checkbox");
        this.hasRadiobutton = types.includes("radio");
        if (this.hasCheckbox) {
            this.updateCheckboxChildren();
        }
    },
    updated(): void {
        if (this.hasCheckbox) {
            this.debouncedUpdateChildren();
        }
    },
    methods: {
        async onValidity({ detail }: CustomEvent<ValidityEvent>): Promise<void> {
            if (detail.target !== this.$el) {
                return;
            }
            this.detail = detail;

            await this.$nextTick();

            const errorMessage = renderSlotText(this.$slots.label) ?? "";

            const firstFocusableElement = this.$el.querySelector(
                "input:not(disabled), select:not(disabled), textarea:not(disabled)",
            );
            const focusElementId = firstFocusableElement ? firstFocusableElement.id : this.id;

            this.validityElement = this.$el as HTMLElement;
            this.dispatchObject = {
                ...detail,
                errorMessage,
                focusElementId,
            };
            this.validity = this.detail;

            if (this.validityElement) {
                dispatchComponentValidityEvent(this.validityElement, this.dispatchObject);
            }

            const message = detail.validityMode === "INITIAL" ? "" : detail.validationMessage;

            if (message !== this.oldMessage) {
                this.forceLegendUpdate();
                this.oldMessage = message;
            }
        },
        /**
         * Workaround for NVDA-bug. Force re rendering of legend element due to NVDA not recognizing innerHTML changes.
         * NVDA has closed the bug as it is related to the browser (works in FF): https://github.com/nvaccess/nvda/issues/13162
         */
        forceLegendUpdate(): void {
            this.legendKey++;
        },
        async updateCheckboxChildren(): Promise<void> {
            await this.$nextTick();
            const checkboxes = Array.from(this.$el.querySelectorAll('input[type="checkbox"]')) as HTMLInputElement[];
            if (!isEqual(this.children, checkboxes)) {
                this.children = checkboxes;
            }
        },
    },
});
</script>

<template>
    <fieldset :id="id" class="fieldset" :class="classes" @validity="onValidity">
        <legend :key="legendKey" class="label" :class="legendClass">
            <!-- @slot Slot for label content. This slot is required. -->
            <slot name="label"></slot>

            <span v-if="hasCheckbox && children.length > 1" class="sr-only">
                <span>{{ numberOfCheckboxesScreenReaderText }}</span>
            </span>
            <!--
                @slot Optional slot for description. See {@link FLabel} for details.
                @binding {string[]} description-class CSS classes for primary description content.
                @binding {string[]} format-description-class CSS classes for format description.
            -->
            <slot name="description" :description-class :format-description-class></slot>

            <slot name="error-message" v-bind="{ hasError, validationMessage: validity.validationMessage }">
                <template v-if="hasError">
                    <span class="label__message label__message--error">
                        <f-icon class="label__icon--left" name="error"></f-icon>
                        {{ validity.validationMessage }}
                    </span>
                </template>
            </slot>
        </legend>

        <span v-if="hasCheckbox" data-test="checked-boxes" class="sr-only" aria-live="polite">
            <span v-if="children.length === 1">{{ checkboxCheckedScreenReaderText }}</span>
            <span v-else>{{ numberOfCheckedCheckboxesScreenText }}</span>
        </span>

        <!-- the original <legend> element is sr-only when a tooltip is present
        so the tooltip button can be positioned correctly when a description is
        also present -->
        <template v-if="hasTooltipSlot">
            <div ref="tooltipAttachTo" class="label">
                <span aria-hidden="true">
                    <slot name="label"></slot>
                </span>
            </div>

            <!-- @slot Slot for tooltip. -->
            <slot name="tooltip"></slot>

            <div
                v-if="hasDescriptionSlot || hasErrorMessageSlot || hasError"
                class="label"
                aria-hidden="true"
                :class="groupLabelClass"
            >
                <!--
                    @slot Optional slot for description. See {@link FLabel} for details.
                    @binding {string[]} description-class CSS classes for primary description content.
                    @binding {string[]} format-description-class CSS classes for format description.
                -->
                <slot name="description" :description-class :format-description-class></slot>

                <!--
                    @slot Slot for displaying single or several error messages.
                    @binding {boolean} hasError Set to true when a validation error is present
                    @binding {string} validationMessage Descriptive validation error message for current error
                -->
                <slot name="error-message" v-bind="{ hasError, validationMessage: validity.validationMessage }">
                    <template v-if="hasError">
                        <span class="label__message label__message--error">
                            <f-icon class="label__icon--left" name="error"></f-icon>
                            {{ validity.validationMessage }}
                        </span>
                    </template>
                </slot>
            </div>
        </template>

        <div :class="groupContentClass">
            <!-- @slot Slot for fieldset content. -->
            <slot name="default"></slot>
        </div>
    </fieldset>
</template>
