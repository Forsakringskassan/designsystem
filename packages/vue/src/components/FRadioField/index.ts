export { default as FRadioField } from "./FRadioField.vue";
