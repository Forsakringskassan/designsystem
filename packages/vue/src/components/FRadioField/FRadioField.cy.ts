import { type DefineComponent, defineComponent } from "vue";
import { ValidationService } from "@fkui/logic";
import {
    DensityWrapper,
    densityWrapperHeight,
    densityWrapperWidth,
} from "@fkui/test-utils/vue";
import { FFieldsetPageObject } from "../../cypress";
import { FTooltip } from "../FTooltip";
import FFieldset from "../FFieldset/FFieldset.vue";
import FRadioField from "./FRadioField.vue";

function createComponent(template: string): DefineComponent {
    return defineComponent({
        template,
        components: {
            FFieldset,
            FRadioField,
            FTooltip,
        },
        data() {
            return {
                radioModel: "",
            };
        },
        mounted() {
            ValidationService.setSubmitted("radio-name");
        },
    });
}

const radioField = {
    yes: ".radio-button:nth(0)",
    no: ".radio-button:nth(1)",
    disabled1: ".radio-button:nth(2)",
    disabled2: ".radio-button:nth(3)",
};

describe("FRadioField", () => {
    const radioGroup = new FFieldsetPageObject("fieldset");

    it("should have label with description", () => {
        const template = /* HTML */ `
            <f-fieldset name="radio-name">
                <template #label> Label text </template>
                <template #description="{ descriptionClass }">
                    <span :class="descriptionClass"> Description text </span>
                </template>
                <f-radio-field v-model="radioModel" :value="true">
                    Yes
                </f-radio-field>
                <f-radio-field v-model="radioModel" :value="false">
                    No
                </f-radio-field>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));
        radioGroup.label.el().should("contain.text", "Label text");
        radioGroup.label
            .description()
            .should("contain.text", "Description text");
    });

    it("should not show error message if not required and submitted", () => {
        const template = /* HTML */ `
            <f-fieldset id="radio-name" name="radio-name">
                <template #label> Label text </template>
                <f-radio-field v-model="radioModel" :value="true">
                    Yes
                </f-radio-field>
                <f-radio-field v-model="radioModel" :value="false">
                    No
                </f-radio-field>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));
        const radioGroupField = radioGroup.radioButton(radioField.yes);
        radioGroup.radioButton(radioField.yes).isSelected().should("be.false");
        radioGroup.radioButton(radioField.no).isSelected().should("be.false");
        radioGroupField.radioButton().focus().blur();
        radioGroup.el().click();
        radioGroup.label.errorMessage().should("not.exist");
    });

    it("should show error message when required and submitted", () => {
        const template = /* HTML */ `
            <f-fieldset id="radio-name" name="radio-name" v-validation.required>
                <template #label> Label text </template>
                <f-radio-field v-model="radioModel" :value="true">
                    Yes
                </f-radio-field>
                <f-radio-field v-model="radioModel" :value="false">
                    No
                </f-radio-field>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));
        const radioGroupField = radioGroup.radioButton(radioField.yes);
        radioGroup.radioButton(radioField.yes).isSelected().should("be.false");
        radioGroup.radioButton(radioField.no).isSelected().should("be.false");
        radioGroupField.radioButton().focus().blur();
        radioGroup.el().click();
        radioGroup.label.errorMessage().should("exist");
        radioGroup.label
            .errorMessage()
            .should("contain.text", "VÃ¤lj ett av alternativen.");
    });

    it("should have fields with text", () => {
        const template = /* HTML */ `
            <f-fieldset name="radio-name">
                <template #label> Label text </template>
                <f-radio-field v-model="radioModel" :value="true">
                    Yes
                </f-radio-field>
                <f-radio-field v-model="radioModel" :value="false">
                    No
                </f-radio-field>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));

        const radioYes = radioGroup.radioButton(radioField.yes);
        const radioNo = radioGroup.radioButton(radioField.no);

        radioGroup.numberOfRadio().should("be.equal", 2);
        radioYes.el().should("contain.text", "Yes");
        radioNo.el().should("contain.text", "No");
    });

    it("should only be able to choose one", () => {
        const template = /* HTML */ `
            <f-fieldset name="radio-name">
                <template #label> Label text </template>
                <f-radio-field v-model="radioModel" :value="true">
                    Yes
                </f-radio-field>
                <f-radio-field v-model="radioModel" :value="false">
                    No
                </f-radio-field>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));

        radioGroup.radioButton(radioField.no).select();
        radioGroup.radioButton(radioField.no).isSelected().should("be.true");
        radioGroup.radioButton(radioField.yes).isSelected().should("be.false");

        radioGroup.radioButton(radioField.yes).select();
        radioGroup.radioButton(radioField.yes).isSelected().should("be.true");
        radioGroup.radioButton(radioField.no).isSelected().should("be.false");
    });

    it("should have tooltip", () => {
        const template = /* HTML */ `
            <f-fieldset name="radio-name">
                <template #label> Label text </template>
                <template #tooltip>
                    <f-tooltip
                        screen-reader-text="Screen reader text."
                        header-tag="h1"
                    >
                        <template #header> Tooltip header </template>
                        <template #body> Tooltip body </template>
                    </f-tooltip>
                </template>
                <f-radio-field v-model="radioModel" :value="true">
                    Yes
                </f-radio-field>
                <f-radio-field v-model="radioModel" :value="false">
                    No
                </f-radio-field>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));
        radioGroup.tooltip.iButton().should("exist");
    });

    it("should not be able to select disabled fields, expanded should not exist", () => {
        const template = /* HTML */ `
            <f-fieldset name="radio-name">
                <template #label> Label text </template>
                <f-radio-field v-model="radioModel" :value="true">
                    Yes
                </f-radio-field>
                <f-radio-field v-model="radioModel" :value="false">
                    No
                </f-radio-field>
                <f-radio-field v-model="radioModel" :value="false" disabled>
                    Disabled 1
                </f-radio-field>
                <f-radio-field v-model="radioModel" :value="false" disabled>
                    Disabled 2
                </f-radio-field>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));
        radioGroup.radioButton(radioField.disabled1).select();
        radioGroup
            .radioButton(radioField.disabled1)
            .isSelected()
            .should("be.false");

        radioGroup.radioButton(radioField.disabled2).select();
        radioGroup
            .radioButton(radioField.disabled2)
            .isSelected()
            .should("be.false");

        radioGroup.radioButton(radioField.yes).select();
        radioGroup.radioButton(radioField.yes).isSelected().should("be.true");
        radioGroup.radioButton(radioField.yes).select();
        //expanded should not exist
        radioGroup.radioButton(radioField.yes).details().should("not.exist");
        radioGroup.radioButton(radioField.no).select();
        //expanded should not exist
        radioGroup.radioButton(radioField.no).details().should("not.exist");
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    it.skip("should always show details regardless of whether yes or no is checked when show-details is always", () => {
        const template = /* HTML */ `
            <f-fieldset name="radio-name" border show-details="always">
                <template #label> Label text </template>
                <template #default>
                    <f-radio-field v-model="radioModel" :value="true">
                        <template #default> Yes </template>
                        <template #details>
                            Value true - Details is always visible
                        </template>
                    </f-radio-field>
                    <f-radio-field v-model="radioModel" :value="false">
                        <template #default> No </template>
                        <template #details>
                            Value false - Details is always visible
                        </template>
                    </f-radio-field>
                </template>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));

        const radio1 = radioGroup.radioButton(radioField.yes);
        const radio2 = radioGroup.radioButton(radioField.no);
        radio1.select();
        radio1.details().should("be.visible");
        radio2.details().should("be.visible");
        cy.get("fieldset").should("exist");
        cy.get("fieldset").toMatchScreenshot({ baseDelay: 500 });
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    it.skip("details should be visible only when item is selected when show-details is when-selected", () => {
        const template = /* HTML */ `
            <f-fieldset name="radio-name" border show-details="when-selected">
                <template #label> Label text </template>
                <template #default>
                    <f-radio-field v-model="radioModel" :value="true">
                        <template #default> Yes </template>
                        <template #details>
                            Value YES - Details is visible only when selected
                        </template>
                    </f-radio-field>
                    <f-radio-field v-model="radioModel" :value="false">
                        <template #default> No </template>
                        <template #details>
                            Value NO - Details is visible only when selected
                        </template>
                    </f-radio-field>
                </template>
            </f-fieldset>
        `;
        cy.mount(createComponent(template));

        const radio1 = radioGroup.radioButton(radioField.yes);
        const radio2 = radioGroup.radioButton(radioField.no);

        radio1.select();
        radio1.details().should("be.visible");
        radio2.details().should("not.exist");
        radio2.select();
        radio1.details().should("not.exist");
        radio2.details().should("be.visible");
        cy.get("fieldset").should("exist");
        cy.get("fieldset").toMatchScreenshot({ baseDelay: 500 });
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    describe.skip("density", () => {
        const DensityComponent = defineComponent({
            template: /* HTML */ `
                <density-wrapper>
                    <f-fieldset :name="fieldset">
                        <template #label>
                            <h2>Radioknappsgrupp</h2>
                        </template>
                        <f-radio-field v-model="radioField" value="Radio1">
                            Text
                        </f-radio-field>
                        <f-radio-field v-model="radioField" value="Radio2">
                            Text
                        </f-radio-field>
                        <f-radio-field v-model="radioField" value="Radio3">
                            Text
                        </f-radio-field>
                        <f-radio-field v-model="radioField" value="Radio4">
                            Text
                        </f-radio-field>
                    </f-fieldset>
                </density-wrapper>
            `,
            components: {
                DensityWrapper,
                FFieldset,
                FRadioField,
            },
            data() {
                return {
                    radioField: "",
                };
            },
        });

        it(`should be densified`, () => {
            cy.viewport(densityWrapperWidth, densityWrapperHeight);
            cy.mount(DensityComponent);
            cy.toMatchScreenshot();
        });
    });
});
