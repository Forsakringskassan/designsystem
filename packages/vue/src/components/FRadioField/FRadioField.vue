<script lang="ts">
import { defineComponent } from "vue";
import { ElementIdService, type ValidityEvent } from "@fkui/logic";
import { useFieldset } from "../FFieldset";
import { renderSlotText, dispatchComponentValidityEvent, hasSlot, getHTMLElementFromVueRef } from "../../utils";

// ! Boolean must be last so that empty string ('') is not parsed as true, see: https://github.com/vuejs/vue/blob/81e1e47cabbd479e2a285f03120944f1efffe749/test/unit/features/options/props.spec.js#L516.
const anyType = [String, Object, Array, Number, Date, Boolean];

export default defineComponent({
    name: "FRadioField",

    inheritAttrs: false,
    props: {
        /**
         * Set to `true`, empty string `""` or string `"disabled"` to disable this input field.
         */
        disabled: {
            type: Boolean,
            required: false,
            default: false,
        },
        /**
         * The id for the input id attribute.
         * The id for the label for attribute.
         * If the prop is not set a random value will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
        /**
         * The value for the input checked attribute.
         * @model
         */
        // ? The rule is disabled so that the `checked` prop can be undefined or null.
        /* eslint-disable-next-line vue/require-default-prop -- technical debt,
        /* it should contain a default value of undefined and proptype should
        /* include undefined (see comment on line above) */
        modelValue: {
            type: anyType,
            required: false,
        },
        /**
         * The value for the input.
         */
        value: {
            type: anyType,
            required: true,
        },
    },
    emits: ["change", "update:modelValue"],
    setup() {
        const { sharedName, showDetails, getFieldsetLabelText } = useFieldset();
        return { sharedName, showDetails, getFieldsetLabelText };
    },
    data() {
        return {
            height: 0,
            initialStyle: {
                overflow: "hidden",
                transition: "height 400ms cubic-bezier(0.46, 0.03, 0.52, 0.96)",
            },
            hiddenStyle: {
                height: "auto",
                position: "absolute",
                visibility: "hidden",
            },
            visibleStyle: {
                width: "",
                position: "",
                visibility: "",
                height: "0px",
            },
            openedStyle: {
                height: "auto",
            },
        };
    },
    computed: {
        attrs(): Record<string, unknown> {
            return {
                ...this.$attrs,
                value: this.value,
                checked: this.value === this.modelValue,
                name: this.sharedName ?? this.$attrs.name,
                onChange: (event: Event) => {
                    if (event.target instanceof HTMLInputElement) {
                        /**
                         * V-model event.
                         * @event update:modelValue
                         * @type {anyType}
                         */
                        this.$emit("update:modelValue", this.value);

                        /**
                         * Emitted when the value of the radiobutton changes.
                         *
                         * @event change
                         * @type {anyType}
                         */
                        this.$emit("change", this.value);
                    }
                },
                onInput: (event: Event) => {
                    // ? Force focus on Mac + Safari
                    (event.target as HTMLInputElement).focus();
                },
            };
        },
        disabledClass(): string {
            return this.disabled ? "disabled" : "";
        },
    },
    methods: {
        async onValidity({ detail }: CustomEvent<ValidityEvent>): Promise<void> {
            if (detail.target !== this.$el.querySelector("input")) {
                return;
            }

            await this.$nextTick();

            let errorMessage = "";

            if (hasSlot(this, "default")) {
                const labelText = this.getFieldsetLabelText();
                if (labelText) {
                    errorMessage = `${labelText} ${renderSlotText(this.$slots.default)}`;
                } else {
                    errorMessage = `${renderSlotText(this.$slots.default)}`;
                }
            }

            const element = this.$el.querySelector(`#${detail.elementId}`);
            if (element) {
                dispatchComponentValidityEvent(element, {
                    ...detail,
                    errorMessage,
                    focusElementId: detail.elementId,
                });
            }
        },
        enter(element: Element): void {
            const htmlElement = getHTMLElementFromVueRef(element);
            const computedStyle = getComputedStyle(element);
            Object.assign(htmlElement.style, this.initialStyle);
            Object.assign(htmlElement.style, this.hiddenStyle);
            htmlElement.style.width = computedStyle.width;
            const height = computedStyle.height;
            Object.assign(htmlElement.style, this.visibleStyle);
            // Force redraw
            /* eslint-disable-next-line @typescript-eslint/no-unused-expressions -- technical debt, there should be a better way */
            getComputedStyle(element).height;
            setTimeout(() => {
                this.height = parseInt(height as string, 10);
                htmlElement.style.height = height;
            });
        },
        afterEnter(element: Element): void {
            const htmlElement = getHTMLElementFromVueRef(element);
            Object.assign(htmlElement.style, this.openedStyle);
        },
        leave(element: Element): void {
            const htmlElement = getHTMLElementFromVueRef(element);
            const height = getComputedStyle(element).height;
            htmlElement.style.height = height;

            // Force redraw
            /* eslint-disable-next-line @typescript-eslint/no-unused-expressions -- technical debt, there should be a better way */
            getComputedStyle(element).height;
            setTimeout(() => {
                Object.assign(htmlElement.style, this.visibleStyle);
            });
        },
    },
});
</script>

<template>
    <div class="radio-button" :class="disabledClass" @validity="onValidity">
        <input :id="id" type="radio" class="radio-button__input" :disabled="disabled" v-bind="attrs" />

        <label :class="$slots.details ? 'radio-button__label radio-button__width' : 'radio-button__label'" :for="id">
            <!-- @slot Slot for label content. -->
            <slot></slot>
            <template v-if="$slots.details">
                <span v-if="showDetails === 'always'" class="radio-button__details">
                    <br />
                    <!-- @slot Slot for details, should only contain short text -->
                    <slot name="details"></slot>
                </span>
                <transition
                    v-if="showDetails === 'when-selected'"
                    @enter="enter"
                    @after-enter="afterEnter"
                    @leave="leave"
                >
                    <span v-if="value === modelValue" class="radio-button__details">
                        <br />
                        <!-- @slot Slot for details, should only contain short text-->
                        <slot name="details" :height="height"></slot>
                    </span>
                </transition>
            </template>
        </label>
    </div>
</template>
