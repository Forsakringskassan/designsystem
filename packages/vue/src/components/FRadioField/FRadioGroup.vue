<template>
    <f-fieldset :id="id" :name="name" v-bind="$attrs" :horizontal="isHorizontal">
        <template #label>
            <!-- @slot Slot for label content. This slot is required. -->
            <slot name="label"></slot>
        </template>

        <template v-if="$slots.tooltip" #tooltip>
            <!-- @slot Slot for tooltip. -->
            <slot name="tooltip"></slot>
        </template>

        <template #description="{ descriptionClass, discreteDescriptionClass }">
            <!--
                @slot Optional slot for description. See [FLabel](#/Components/FLabel) for details.
                @binding {string[]} descriptionClass CSS classes for primary description content.
                @binding {string[]} discreteDescriptionClass CSS classes for format description.
            -->
            <slot name="description" v-bind="{ descriptionClass, discreteDescriptionClass }"></slot>
        </template>

        <template #error-message>
            <!-- @slot Slot for displaying single or several error messages. -->
            <slot name="error-message"></slot>
        </template>

        <!-- @slot
Slot for main content, usually radio buttons.

Slot content is available through `v-slot="{ <propertyName> }"`, e.g.
`v-slot="{ indentClass }"`.

The following properties are available:
* `indentClass: string[];` CSS class to use for nesting content under a radio button. Use with `:class="indentClass"`.
        -->
        <template #default>
            <slot v-bind="{ indentClass: 'indent' }" name="default"></slot>
        </template>
    </f-fieldset>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { ElementIdService } from "@fkui/logic";
import { FFieldset } from "../FFieldset";

/**
 * @deprecated `<f-radio-group>` is deprecated.
 * See migration guide: https://forsakringskassan.github.io/latest/guide/migration/migrating-to-fieldset.html
 */
export default defineComponent({
    name: "FRadioGroup",
    components: { FFieldset },
    inheritAttrs: false,
    props: {
        /**
         * The id for the fieldset id attribute.
         * If the prop is not set the id will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
        /**
         * The name of the radio group.
         * The radio group fields in the group will use the same name.
         */
        name: {
            type: String,
            required: true,
        },
        /**
         * If radio buttons should be aligned horizontally.
         * If the prop is not set the radio buttons will be aligned vertically.
         * Default: `false`
         */
        isHorizontal: {
            type: Boolean,
            required: false,
        },
    },
});
</script>
