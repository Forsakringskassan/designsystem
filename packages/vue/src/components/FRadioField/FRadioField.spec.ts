import { VueWrapper, mount } from "@vue/test-utils";
import { createPlaceholderInDocument } from "@fkui/test-utils/vue";
import { injectionKeys as fieldsetInjectionKeys } from "../FFieldset/use-fieldset";
import FRadioField from "./FRadioField.vue";

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
} = {}): VueWrapper {
    return mount(FRadioField, {
        attrs: { ...attrs },
        attachTo: createPlaceholderInDocument(),
        props: { value: "Default value", ...props },
        slots: { default: "Default label", ...slots },
        global: {
            provide: {
                [fieldsetInjectionKeys.sharedName as symbol]: "providedName",
            },
        },
    });
}

it.each`
    vModel       | value    | expected
    ${undefined} | ${false} | ${false}
    ${null}      | ${false} | ${false}
    ${""}        | ${false} | ${false}
    ${false}     | ${""}    | ${false}
    ${0}         | ${false} | ${false}
    ${undefined} | ${true}  | ${false}
    ${null}      | ${true}  | ${false}
    ${""}        | ${true}  | ${false}
    ${true}      | ${""}    | ${false}
    ${0}         | ${true}  | ${false}
    ${0}         | ${0}     | ${true}
    ${1}         | ${0}     | ${false}
    ${1}         | ${1}     | ${true}
    ${0}         | ${1}     | ${false}
`(
    'should handle v-model value "$vModel" where checked should be "$expected" when radio value is "$value"',
    ({ vModel, value, expected }) => {
        const wrapper = createWrapper({
            props: {
                value,
                modelValue: vModel,
            },
        });

        expect(wrapper.get("input").element.checked).toBe(expected);
    },
);

describe("snapshots", () => {
    it("should match snapshot with label and input", () => {
        const wrapper = createWrapper();
        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("attributes", () => {
    it("should pass attributes", () => {
        const wrapper = createWrapper({
            attrs: {
                disabled: true,
                required: true,
            },
        });
        const input = wrapper.get("input");

        expect(input.attributes("disabled")).toBeDefined();
        expect(input.attributes("required")).toBeDefined();
    });
});

describe("disabled", () => {
    it.each`
        disabled      | expectedResult | description
        ${"disabled"} | ${true}        | ${"be"}
        ${true}       | ${true}        | ${"be"}
        ${false}      | ${false}       | ${"not be"}
    `(
        "should $description disabled when disabled prop is $disabledAttribute",
        async ({ disabled, expectedResult }) => {
            const wrapper = createWrapper({
                props: {
                    disabled,
                },
            });
            const input = wrapper.get("input").element;
            expect(input.disabled).toBe(expectedResult);
            expect(wrapper.classes("disabled")).toBe(expectedResult);
        },
    );
});

describe("events", () => {
    it("should support v-model by emitting update:modelValue event with value", async () => {
        const wrapper = createWrapper({
            props: { value: "Some value", modelValue: "Some value" },
        });

        const input = wrapper.get("input");

        const htmlInput = input.element;

        expect(htmlInput.checked).toBe(true);
        await wrapper.setProps({ modelValue: undefined });
        expect(htmlInput.checked).toBe(false);

        await input.trigger("click");

        expect(
            wrapper.emitted("update:modelValue")![0][0],
        ).toMatchInlineSnapshot(`"Some value"`);
    });

    it("should emit change event when input value changes", async () => {
        const wrapper = createWrapper({
            props: { value: true, modelValue: false },
        });

        const input = wrapper.get("input");

        await input.trigger("click");
        expect(wrapper.emitted("change")![0]).toMatchInlineSnapshot(`
            [
              true,
            ]
        `);
    });

    it("should pass listeners", async () => {
        const foobar = jest.fn();
        const wrapper = createWrapper({
            attrs: { onFoobar: foobar },
        });
        const element = wrapper.get("input");
        await element.trigger("foobar");
        expect(foobar).toHaveBeenCalled();
    });

    it("should pass click listener and trigger focus on click", async () => {
        const click = jest.fn();

        const wrapper = createWrapper({
            attrs: { onClick: click },
        });

        const input = wrapper.get("input");
        const htmlInput = input.element;
        htmlInput.focus = jest.fn();

        await input.trigger("click");
        expect(click).toHaveBeenCalled();
        expect(htmlInput.focus).toHaveBeenCalled();
    });
});
