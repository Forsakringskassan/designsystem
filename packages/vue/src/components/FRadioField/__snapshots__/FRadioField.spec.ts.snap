// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot with label and input 1`] = `
<div
  class="radio-button"
>
  <input
    class="radio-button__input"
    id="fkui-vue-element-0001"
    name="providedName"
    type="radio"
    value="Default value"
  />
  <label
    class="radio-button__label"
    for="fkui-vue-element-0001"
  >
    
    Default label
    
    <!--v-if-->
  </label>
</div>
`;
